sdk/storage/Azure.Storage.DataMovement/src/DownloadChunkHandler.cs

```diff
@@ -0,0 +1,281 @@
+        {
+            // If there are any other chunks that have already been downloaded that
+            // can be appended to the file, let's do it now.
+            while ((_bytesTransferred < _expectedLength) &&
+                    (_currentRangeIndex < _rangesCount) &&
+                    _rangesCompleted.ContainsKey(_ranges[_currentRangeIndex].Offset))
+            {
+                HttpRange currentRange = _ranges[_currentRangeIndex];
+                if (_rangesCompleted.TryRemove(currentRange.Offset, out string chunkFilePath))
+                {
+                    try
+                    {
+                        if (File.Exists(chunkFilePath))
+                        {
+                            using (Stream content = File.OpenRead(chunkFilePath))
+                            {
+                                await _copyToDestinationFile(currentRange.Offset, (long) currentRange.Length, content).ConfigureAwait(false);
+                            }
+                            // Delete the temporary chunk file that's no longer needed
+                            File.Delete(chunkFilePath);
+                        }
+                        else
+                        {
+                            await InvokeFailedEvent(new FileNotFoundException(
+                                $"Could not append chunk to destination file at Offset: " +
+                                $"\"{currentRange.Offset}\" and Length: \"{currentRange.Length}\"," +
+                                $"due to the chunk file missing: \"{chunkFilePath}\"")).ConfigureAwait(false);
+                        }
+                    }
+                    catch
+                    {
+                        await InvokeFailedEvent(new Exception(
+                            $"Could not append chunk to destination file at Offset: " +
+                            $"\"{currentRange.Offset}\" and Length: \"{currentRange.Length}\""))
+                            .ConfigureAwait(false);
+                    }
+                }
+                else
+                {
+                    await InvokeFailedEvent(new ArgumentOutOfRangeException(
+                        "Offset",
+                        currentRange,
+                        $"Cannot find offset returned by Successful Download Range at Offset: " +
+                        $"\"{currentRange.Offset}\" and Length: \"{currentRange.Length}\""))
+                        .ConfigureAwait(false);
+                }
+
+                // Increment the current range we are expect, if it's null then
+                // that's the next one we have to wait on.
+                UpdateBytesAndRange((long)_ranges[_currentRangeIndex].Length);
+            }
+        }
+
+        private async Task InvokeFailedEvent(Exception ex)
+        {
+            foreach (HttpRange range in _ranges)
+            {
+                if (_rangesCompleted.TryRemove(range.Offset, out string tempChunkFile))
```

Comment:
Is it possible we fail while some chunks are being written and not in `_rangesComplete`

Reply:
It's possible that we could fail in trying to removing the file, because it's being written. I think at that point we can't stop it in flight and will be removed anyways after it's finished being written.